sql: panic: cannot overwrite value on flat Bytes

```
CREATE TABLE IF NOT EXISTS tab_orig AS
        SELECT g FROM generate_series(0, 5) AS g;

SET vectorize = experimental_on;

SELECT
        ''
FROM
        tab_orig JOIN tab_orig AS tab_2054 ON true;
```

```
panic: cannot overwrite value on flat Bytes: maxSetIndex=35, setIndex=0, consider using Reset [recovered]
        panic: cannot overwrite value on flat Bytes: maxSetIndex=35, setIndex=0, consider using Reset [recovered]
        panic: panic while executing 1 statements: SELECT _ FROM _ JOIN _ AS _ ON _; caused by cannot overwrite value on flat Bytes: maxSetIndex=35, setIndex=0, consider using Reset

goroutine 1135 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc001860880, 0x4bd54c0, 0xc000d8d9c0, 0x3c55880, 0xc001986a60)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc001860880, 0x4bd54c0, 0xc000d8d9c0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61
panic(0x3c55880, 0xc001986a60)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/exec/execerror.CatchVectorizedRuntimeError.func1(0xc00180c1c8)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/execerror/error.go:82 +0x720
panic(0x3c55880, 0xc001986a60)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/col/coldata.(*Bytes).Set(0xc00193f360, 0x0, 0x0, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/col/coldata/bytes.go:61 +0x274
github.com/cockroachdb/cockroach/pkg/sql/exec.constBytesOp.Next(0x4bd8440, 0xc001331590, 0x1, 0x0, 0x0, 0x0, 0x0, 0x4bd54c0, 0xc00184e700, 0x7f1391484c00, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/const.eg.go:163 +0x239
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).next(0xc000b10700, 0x726aef, 0x8, 0xc001510c80, 0x4bf2300)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:144 +0x40e
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).nextAdapter(...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:135
github.com/cockroachdb/cockroach/pkg/sql/exec/execerror.CatchVectorizedRuntimeError(0xc00180c1d8, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/execerror/error.go:91 +0x55
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).Next(0xc000b10700, 0xc00184e700, 0x4264aa5, 0xc, 0x4bd54c0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:179 +0x4d
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x4bd54c0, 0xc00184e700, 0x4bfba80, 0xc000b10700, 0x4bae180, 0xc0019402c0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc000b10700, 0x4bd54c0, 0xc00184e700)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc001a257a0, 0x4bd5580, 0xc00196b3b0, 0x43d99f8, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:584 +0x1ec
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000110640, 0xc001896f60, 0xc0014a73b0, 0xc00180c878, 0xc0019402c0, 0xc001860ca0, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:334 +0x381
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc000110640, 0x4bd5580, 0xc001866960, 0xc001860ca0, 0xc001896f60, 0xc0014a73b0, 0x4bd7300, 0xc001a25560, 0xc0019402c0, 0xc001860a58)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:959 +0x20a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc001860880, 0x4bd5580, 0xc001866960, 0xc001860bd0, 0x3, 0x7f1393a17328, 0xc0018f2f68, 0xc00180ca01, 0x0, 0x0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:808 +0x370
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc001860880, 0x4bd5580, 0xc001866960, 0xc001860bd0, 0x7f1393a17328, 0xc0018f2f68, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:700 +0x75b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc001860880, 0x4bd5580, 0xc001866960, 0x4bf5300, 0xc0013a1380, 0xc0011f2330, 0x49, 0x0, 0x2, 0x0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:416 +0xb64
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc001860880, 0x4bd5580, 0xc001866960, 0x4bf5300, 0xc0013a1380, 0xc0011f2330, 0x49, 0x0, 0x2, 0x0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:98 +0x4ec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc001860880, 0x4bd54c0, 0xc0013a13c0, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1211 +0x1aa2
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc001860880, 0x4bd54c0, 0xc000d8d9c0, 0xc000538a00, 0x5400, 0x15000, 0xc000538a98, 0xc00043bc00, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1140 +0x1a3
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000133680, 0x4bd54c0, 0xc000d8d9c0, 0xc001860880, 0x5400, 0x15000, 0xc000538a98, 0xc00043bc00, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc00048cc1f, 0xc000432980, 0x4bd54c0, 0xc000d8d9c0, 0xc00043bc00, 0xc000133680, 0xc0018f2900, 0x4bf3880, 0xc000432960, 0xc000a2bd40, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:584 +0x21c
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:513 +0x17b
```

See #39827

--- a/pkg/col/coldata/bytes.go
+++ b/pkg/col/coldata/bytes.go
@@ -15,8 +15,84 @@ import (
 	"strings"
 )
 
-// Bytes is a wrapper type for a two-dimensional byte slice ([][]byte).
+// Bytes is a wrapper type for a two-dimensional byte slice.
 type Bytes struct {
+	data [][]byte
+}
+
+// NewBytes returns a Bytes struct with enough capacity for n elements.
+func NewBytes(n int) *Bytes {
+	return &Bytes{
+		data: make([][]byte, n),
+	}
+}
+
+// Get returns the ith []byte in Bytes.
+func (b *Bytes) Get(i int) []byte {
+	return b.data[i]
+}
+
+// Set sets the ith []byte in Bytes.
+func (b *Bytes) Set(i int, v []byte) {
+	b.data[i] = v
+}
+
+// Swap swaps the ith []byte with the jth []byte and vice-versa.
+func (b *Bytes) Swap(i int, j int) {
+	b.data[i], b.data[j] = b.data[j], b.data[i]
+}
+
+// Slice returns a new Bytes struct with its internal data sliced according to
+// start and end.
+func (b *Bytes) Slice(start, end int) *Bytes {
+	return &Bytes{data: b.data[start:end]}
+}
+
+// CopySlice copies srcStartIdx inclusive and srcEndIdx inclusive []byte values
+// from src into the receiver starting at destIdx.
+func (b *Bytes) CopySlice(src *Bytes, destIdx, srcStartIdx, srcEndIdx int) {
+	copy(b.data[destIdx:], src.data[srcStartIdx:srcEndIdx])
+}
+
+// AppendSlice appends srcStartIdx inclusive and srcEndIdx inclusive []byte
+// values from src into the receiver starting at destIdx.
+func (b *Bytes) AppendSlice(src *Bytes, destIdx, srcStartIdx, srcEndIdx int) {
+	b.data = append(b.data[:destIdx], src.data[srcStartIdx:srcEndIdx]...)
+}
+
+// AppendVal appends the given []byte value to the end of the receiver.
+func (b *Bytes) AppendVal(v []byte) {
+	b.data = append(b.data, v)
+}
+
+// Len returns how many []byte values the receiver contains.
+func (b *Bytes) Len() int {
+	return len(b.data)
+}
+
+var zeroBytesColumn = make([][]byte, BatchSize)
+
+// Zero zeroes out the underlying bytes.
+func (b *Bytes) Zero() {
+	for n := 0; n < len(b.data); n += copy(b.data[n:], zeroBytesColumn) {
+	}
+}
+
+// PrimitiveRepr is a temprorary migration tool.
+// TODO(asubiotto): Remove this.
+func (b *Bytes) PrimitiveRepr() [][]byte {
+	return b.data
+}
+
+// Reset here is a stub that is added so that the calls that used to be to
+// flatBytes.Reset() don't need to be removed.
+func (b *Bytes) Reset() {}
+
+// TODO(yuzefovich): fix flatBytes implementation and use that instead of
+// Bytes.
+
+// flatBytes is a wrapper type for a two-dimensional byte slice ([][]byte).
+type flatBytes struct {
 	// data is the slice of all bytes.
 	data []byte
 	// offsets contains the offsets for each []byte slice in data.
@@ -32,11 +108,11 @@ type Bytes struct {
 	maxSetIndex int
 }
 
-// NewBytes returns a Bytes struct with enough capacity for n zero-length
-// []byte values. It is legal to call Set on the returned Bytes at this point,
+// newFlatBytes returns a flatBytes struct with enough capacity for n zero-length
+// []byte values. It is legal to call Set on the returned flatBytes at this point,
 // but Get is undefined until at least one element is Set.
-func NewBytes(n int) *Bytes {
-	return &Bytes{
+func newFlatBytes(n int) *flatBytes {
+	return &flatBytes{
 		// Given that the []byte slices are of variable length, we multiply the
 		// number of elements by some constant factor.
 		// TODO(asubiotto): Make this tunable.
@@ -46,17 +122,17 @@ func NewBytes(n int) *Bytes {
 	}
 }
 
-// Get returns the ith []byte in Bytes. Note that the returned byte slice is
+// Get returns the ith []byte in flatBytes. Note that the returned byte slice is
 // unsafe for reuse if any write operation happens.
-func (b Bytes) Get(i int) []byte {
+func (b *flatBytes) Get(i int) []byte {
 	return b.data[b.offsets[i] : b.offsets[i]+b.lengths[i]]
 }
 
-// Set sets the ith []byte in Bytes. Overwriting a value that is not at the end
-// of the Bytes is not allowed since it complicates memory movement to make/take
+// Set sets the ith []byte in flatBytes. Overwriting a value that is not at the end
+// of the flatBytes is not allowed since it complicates memory movement to make/take
 // away necessary space in the flat buffer. Note that a nil value will be
 // "converted" into an empty byte slice.
-func (b *Bytes) Set(i int, v []byte) {
+func (b *flatBytes) Set(i int, v []byte) {
 	if i < b.maxSetIndex {
 		panic(
 			fmt.Sprintf(
@@ -79,16 +155,16 @@ func (b *Bytes) Set(i int, v []byte) {
 
 // Swap swaps the ith []byte with the jth []byte and vice-versa. This is
 // disallowed.
-func (b *Bytes) Swap(i int, j int) {
+func (b *flatBytes) Swap(i int, j int) {
 	panic("cannot swap flat Bytes elements")
 }
 
-// Slice modifies and returns the receiver Bytes struct sliced according to
+// Slice modifies and returns the receiver flatBytes struct sliced according to
 // start and end. Returning the result is not necessary but is done for
 // compatibility with the usage of other data representations in vectorized
 // execution. Note that Slicing can be expensive since it incurs a step to
 // translate offsets.
-func (b *Bytes) Slice(start, end int) *Bytes {
+func (b *flatBytes) Slice(start, end int) *flatBytes {
 	if start == 0 && end == len(b.offsets) {
 		return b
 	}
@@ -115,14 +191,14 @@ func (b *Bytes) Slice(start, end int) *Bytes {
 // min(dest.Len(), src.Len()) values will be copied. Note that if the length of
 // the receiver is greater than the length of the source, bytes will have to be
 // physically moved. Consider the following example:
-// dest Bytes: "helloworld", offsets: []int32{0, 5}, lengths: []int32{5, 5}
-// src Bytes: "a", offsets: []int32{0}, lengths: []int32{1}
+// dest flatBytes: "helloworld", offsets: []int32{0, 5}, lengths: []int32{5, 5}
+// src flatBytes: "a", offsets: []int32{0}, lengths: []int32{1}
 // If we copy src into the beginning of dest, we will have to move "world" so
 // that the result is:
-// result Bytes: "aworld", offsets: []int32{0, 1}, lengths: []int32{1, 5}
+// result flatBytes: "aworld", offsets: []int32{0, 1}, lengths: []int32{1, 5}
 // Similarly, if "a", is instead "alongerstring", "world" would have to be
 // shifted right.
-func (b *Bytes) CopySlice(src *Bytes, destIdx, srcStartIdx, srcEndIdx int) {
+func (b *flatBytes) CopySlice(src *flatBytes, destIdx, srcStartIdx, srcEndIdx int) {
 	if destIdx < 0 || destIdx > b.Len() {
 		panic(fmt.Sprintf("dest index %d out of range (len=%d)", destIdx, b.Len()))
 	} else if srcStartIdx < 0 || srcStartIdx >= src.Len() || srcEndIdx > src.Len() || srcStartIdx > srcEndIdx {
@@ -202,7 +278,7 @@ func (b *Bytes) CopySlice(src *Bytes, destIdx, srcStartIdx, srcEndIdx int) {
 
 // AppendSlice appends srcStartIdx inclusive and srcEndIdx exclusive []byte
 // values from src into the receiver starting at destIdx.
-func (b *Bytes) AppendSlice(src *Bytes, destIdx, srcStartIdx, srcEndIdx int) {
+func (b *flatBytes) AppendSlice(src *flatBytes, destIdx, srcStartIdx, srcEndIdx int) {
 	if destIdx < 0 || destIdx > b.Len() {
 		panic(fmt.Sprintf("dest index %d out of range (len=%d)", destIdx, b.Len()))
 	} else if srcStartIdx < 0 || srcStartIdx >= src.Len() || srcEndIdx > src.Len() || srcStartIdx > srcEndIdx {
@@ -230,7 +306,7 @@ func (b *Bytes) AppendSlice(src *Bytes, destIdx, srcStartIdx, srcEndIdx int) {
 	}
 
 	// Calculate the amount to translate offsets by before modifying any
-	// destination offsets since we might be appending from the same Bytes (and
+	// destination offsets since we might be appending from the same flatBytes (and
 	// might be overwriting information).
 	translateBy := destDataIdx - src.offsets[srcStartIdx]
 	b.data = append(b.data[:destDataIdx], srcDataToAppend...)
@@ -250,7 +326,7 @@ func (b *Bytes) AppendSlice(src *Bytes, destIdx, srcStartIdx, srcEndIdx int) {
 
 // AppendVal appends the given []byte value to the end of the receiver. A nil
 // value will be "converted" into an empty byte slice.
-func (b *Bytes) AppendVal(v []byte) {
+func (b *flatBytes) AppendVal(v []byte) {
 	b.maxSetIndex = len(b.offsets)
 	b.offsets = append(b.offsets, int32(len(b.data)))
 	b.lengths = append(b.lengths, int32(len(v)))
@@ -258,7 +334,7 @@ func (b *Bytes) AppendVal(v []byte) {
 }
 
 // Len returns how many []byte values the receiver contains.
-func (b Bytes) Len() int {
+func (b *flatBytes) Len() int {
 	return len(b.offsets)
 }
 
@@ -267,7 +343,7 @@ var zeroInt32Slice = make([]int32, BatchSize)
 // Zero zeroes out the underlying bytes. Note that this doesn't change the
 // length. Use this instead of Reset if you need to be able to Get zeroed byte
 // slices.
-func (b Bytes) Zero() {
+func (b *flatBytes) Zero() {
 	b.data = b.data[:0]
 	for n := 0; n < len(b.offsets); n += copy(b.offsets[n:], zeroInt32Slice) {
 	}
@@ -276,18 +352,18 @@ func (b Bytes) Zero() {
 	b.maxSetIndex = 0
 }
 
-// Reset resets the underlying Bytes for reuse.
+// Reset resets the underlying flatBytes for reuse.
 // This is useful when the caller is going to overwrite the underlying bytes and
 // needs a quick way to Reset. Note that this doesn't change the length either.
 // TODO(asubiotto): Move towards removing Set in favor of AppendVal. At that
 //  point we can reset the length to 0.
-func (b *Bytes) Reset() {
+func (b *flatBytes) Reset() {
 	b.data = b.data[:0]
 	b.maxSetIndex = 0
 }
 
 // String is used for debugging purposes.
-func (b Bytes) String() string {
+func (b *flatBytes) String() string {
 	var builder strings.Builder
 	for i := range b.offsets {
 		builder.WriteString(
@@ -297,6 +373,7 @@ func (b Bytes) String() string {
 	return builder.String()
 }
 
+/*
 // BytesFromArrowSerializationFormat takes an Arrow byte slice and accompanying
 // offsets and populates b.
 func BytesFromArrowSerializationFormat(b *Bytes, data []byte, offsets []int32) {
@@ -317,9 +394,10 @@ func BytesFromArrowSerializationFormat(b *Bytes, data []byte, offsets []int32) {
 	b.maxSetIndex = len(offsets) - 1
 }
 
+
 // ToArrowSerializationFormat returns a bytes slice and offsets that are
 // Arrow-compatible. n is the number of elements to serialize.
-func (b Bytes) ToArrowSerializationFormat(n int) ([]byte, []int32) {
+func (b *Bytes) ToArrowSerializationFormat(n int) ([]byte, []int32) {
 	if n == 0 {
 		return []byte{}, []int32{0}
 	}
@@ -328,3 +406,4 @@ func (b Bytes) ToArrowSerializationFormat(n int) ([]byte, []int32) {
 	offsets := append(b.offsets[:n], serializeLength)
 	return data, offsets
 }
+*/

--- a/pkg/col/coldata/bytes_test.go
+++ b/pkg/col/coldata/bytes_test.go
@@ -56,11 +56,11 @@ var bytesMethods = []bytesMethod{set, slice, copySlice, appendSlice, appendVal}
 // [][]byte implementation and checks if the results are equal. If
 // selfReferencingSources is true, this is an indication by the caller that we
 // are testing an edge case where the source for copies/appends refers to the
-// destination. In cases where *Bytes updates itself under the hood, we also
+// destination. In cases where *flatBytes updates itself under the hood, we also
 // update the corresponding b2Source to mirror the behavior.
 func applyMethodsAndVerify(
 	rng *rand.Rand,
-	b1, b1Source *Bytes,
+	b1, b1Source *flatBytes,
 	b2, b2Source [][]byte,
 	methods []bytesMethod,
 	selfReferencingSources bool,
@@ -158,7 +158,7 @@ func applyMethodsAndVerify(
 	return nil
 }
 
-func verifyEqual(flat *Bytes, b [][]byte) error {
+func verifyEqual(flat *flatBytes, b [][]byte) error {
 	if flat.Len() != len(b) {
 		return errors.Errorf("mismatched lengths %d != %d", flat.Len(), len(b))
 	}
@@ -180,7 +180,7 @@ func prettyByteSlice(b [][]byte) string {
 	return builder.String()
 }
 
-func TestBytesRefImpl(t *testing.T) {
+func TestFlatBytesRefImpl(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
 	rng, _ := randutil.NewPseudoRand()
@@ -192,7 +192,7 @@ func TestBytesRefImpl(t *testing.T) {
 
 	n := 1 + rng.Intn(maxLength)
 
-	flat := NewBytes(n)
+	flat := newFlatBytes(n)
 	reference := make([][]byte, n)
 	for i := 0; i < n; i++ {
 		v := make([]byte, rng.Intn(16))
@@ -210,7 +210,7 @@ func TestBytesRefImpl(t *testing.T) {
 	if rng.Float64() < 0.5 {
 		selfReferencingSources = false
 		sourceN = 1 + rng.Intn(maxLength)
-		flatSource = NewBytes(sourceN)
+		flatSource = newFlatBytes(sourceN)
 		referenceSource = make([][]byte, sourceN)
 		for i := 0; i < sourceN; i++ {
 			v := make([]byte, rng.Intn(16))
@@ -238,11 +238,11 @@ func TestBytesRefImpl(t *testing.T) {
 	}
 }
 
-func TestBytes(t *testing.T) {
+func TestFlatBytes(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
 	t.Run("Simple", func(t *testing.T) {
-		b1 := NewBytes(0)
+		b1 := newFlatBytes(0)
 		b1.AppendVal([]byte("hello"))
 		require.Equal(t, "hello", string(b1.Get(0)))
 		b1.AppendVal(nil)
@@ -258,7 +258,7 @@ func TestBytes(t *testing.T) {
 		// However, it is legal to overwrite the last value.
 		b1.Set(1, []byte("ok"))
 
-		// If we Zero the Bytes, we can Set any index.
+		// If we Zero the flatBytes, we can Set any index.
 		b1.Zero()
 		b1.Set(1, []byte("new usage"))
 		// But not an index before that.
@@ -282,8 +282,8 @@ func TestBytes(t *testing.T) {
 	})
 
 	t.Run("Append", func(t *testing.T) {
-		b1 := NewBytes(0)
-		b2 := NewBytes(0)
+		b1 := newFlatBytes(0)
+		b2 := newFlatBytes(0)
 		b2.AppendVal([]byte("source bytes value"))
 		b1.AppendVal([]byte("one"))
 		b1.AppendVal([]byte("two"))
@@ -304,7 +304,7 @@ func TestBytes(t *testing.T) {
 			}
 		}
 
-		b2 = NewBytes(0)
+		b2 = newFlatBytes(0)
 		b2.AppendVal([]byte("hello again"))
 		b2.AppendVal([]byte("hello again"))
 		b2.AppendVal([]byte("hello again"))
@@ -317,8 +317,8 @@ func TestBytes(t *testing.T) {
 	})
 
 	t.Run("Copy", func(t *testing.T) {
-		b1 := NewBytes(0)
-		b2 := NewBytes(0)
+		b1 := newFlatBytes(0)
+		b2 := newFlatBytes(0)
 		b1.AppendVal([]byte("one"))
 		b1.AppendVal([]byte("two"))
 		b1.AppendVal([]byte("three"))

--- a/pkg/col/colserde/arrowbatchconverter.go
+++ b/pkg/col/colserde/arrowbatchconverter.go
@@ -15,6 +15,7 @@ import (
 	"reflect"
 	"unsafe"
 
+	"github.com/apache/arrow/go/arrow"
 	"github.com/apache/arrow/go/arrow/array"
 	"github.com/apache/arrow/go/arrow/memory"
 	"github.com/cockroachdb/cockroach/pkg/col/coldata"
@@ -33,6 +34,9 @@ type ArrowBatchConverter struct {
 	builders struct {
 		// boolBuilder builds arrow bool columns as a bitmap from a bool slice.
 		boolBuilder *array.BooleanBuilder
+		// binaryBuilder builds arrow []byte columns as one []byte slice with
+		// accompanying offsets from a [][]byte slice.
+		binaryBuilder *array.BinaryBuilder
 	}
 
 	scratch struct {
@@ -56,12 +60,13 @@ func NewArrowBatchConverter(typs []coltypes.T) (*ArrowBatchConverter, error) {
 	}
 	c := &ArrowBatchConverter{typs: typs}
 	c.builders.boolBuilder = array.NewBooleanBuilder(memory.DefaultAllocator)
+	c.builders.binaryBuilder = array.NewBinaryBuilder(memory.DefaultAllocator, arrow.BinaryTypes.Binary)
 	c.scratch.arrowData = make([]*array.Data, len(typs))
 	c.scratch.buffers = make([][]*memory.Buffer, len(typs))
 	for i := range c.scratch.buffers {
-		// Most types need only two buffers: one for the nulls, and one for the
-		// values, but some type (i.e. Bytes) need an extra buffer for the offsets.
-		c.scratch.buffers[i] = make([]*memory.Buffer, 0, 3)
+		// Only primitive types are directly constructed, therefore we only need
+		// two buffers: one for the nulls, the other for the values.
+		c.scratch.buffers[i] = make([]*memory.Buffer, 2)
 	}
 	return c, nil
 }
@@ -112,11 +117,20 @@ func (c *ArrowBatchConverter) BatchToArrow(batch coldata.Batch) ([]*array.Data, 
 			arrowBitmap = n.NullBitmap()
 		}
 
-		if typ == coltypes.Bool {
-			// Bools are handled differently from other types. Refer to the comment on
-			// ArrowBatchConverter.builders for more information.
-			c.builders.boolBuilder.AppendValues(vec.Bool()[:n], nil /* valid */)
-			data := c.builders.boolBuilder.NewBooleanArray().Data()
+		if typ == coltypes.Bool || typ == coltypes.Bytes {
+			// Bools and Bytes are handled differently from other coltypes. Refer to the
+			// comment on ArrowBatchConverter.builders for more information.
+			var data *array.Data
+			switch typ {
+			case coltypes.Bool:
+				c.builders.boolBuilder.AppendValues(vec.Bool()[:n], nil /* valid */)
+				data = c.builders.boolBuilder.NewBooleanArray().Data()
+			case coltypes.Bytes:
+				c.builders.binaryBuilder.AppendValues(vec.Bytes().PrimitiveRepr()[:n], nil /* valid */)
+				data = c.builders.binaryBuilder.NewBinaryArray().Data()
+			default:
+				panic(fmt.Sprintf("unexpected type %s", typ))
+			}
 			if arrowBitmap != nil {
 				// Overwrite empty null bitmap with the true bitmap.
 				data.Buffers()[0] = memory.NewBufferBytes(arrowBitmap)
@@ -126,8 +140,8 @@ func (c *ArrowBatchConverter) BatchToArrow(batch coldata.Batch) ([]*array.Data, 
 		}
 
 		var (
-			values  []byte
-			offsets []byte
+			values []byte
+
 			// dataHeader is the reflect.SliceHeader of the coldata.Vec's underlying
 			// data slice that we are casting to bytes.
 			dataHeader *reflect.SliceHeader
@@ -136,15 +150,6 @@ func (c *ArrowBatchConverter) BatchToArrow(batch coldata.Batch) ([]*array.Data, 
 		)
 
 		switch typ {
-		case coltypes.Bytes:
-			var int32Offsets []int32
-			values, int32Offsets = vec.Bytes().ToArrowSerializationFormat(n)
-			// Cast int32Offsets to []byte.
-			int32Header := (*reflect.SliceHeader)(unsafe.Pointer(&int32Offsets))
-			offsetsHeader := (*reflect.SliceHeader)(unsafe.Pointer(&offsets))
-			offsetsHeader.Data = int32Header.Data
-			offsetsHeader.Len = int32Header.Len * sizeOfInt32
-			offsetsHeader.Cap = int32Header.Cap * sizeOfInt32
 		case coltypes.Int8:
 			ints := vec.Int8()[:n]
 			dataHeader = (*reflect.SliceHeader)(unsafe.Pointer(&ints))
@@ -173,22 +178,16 @@ func (c *ArrowBatchConverter) BatchToArrow(batch coldata.Batch) ([]*array.Data, 
 			panic(fmt.Sprintf("unsupported type for conversion to arrow data %s", typ))
 		}
 
-		// Cast values if not set (mostly for non-byte types).
-		if values == nil {
-			valuesHeader := (*reflect.SliceHeader)(unsafe.Pointer(&values))
-			valuesHeader.Data = dataHeader.Data
-			valuesHeader.Len = dataHeader.Len * datumSize
-			valuesHeader.Cap = dataHeader.Cap * datumSize
-		}
+		// Cast values.			// Cast values if not set (mostly for non-byte types).
+		valuesHeader := (*reflect.SliceHeader)(unsafe.Pointer(&values))
+		valuesHeader.Data = dataHeader.Data
+		valuesHeader.Len = dataHeader.Len * datumSize
+		valuesHeader.Cap = dataHeader.Cap * datumSize
 
 		// Construct the underlying arrow buffers.
 		// WARNING: The ordering of construction is critical.
-		c.scratch.buffers[i] = c.scratch.buffers[i][:0]
-		c.scratch.buffers[i] = append(c.scratch.buffers[i], memory.NewBufferBytes(arrowBitmap))
-		if offsets != nil {
-			c.scratch.buffers[i] = append(c.scratch.buffers[i], memory.NewBufferBytes(offsets))
-		}
-		c.scratch.buffers[i] = append(c.scratch.buffers[i], memory.NewBufferBytes(values))
+		c.scratch.buffers[i][0] = memory.NewBufferBytes(arrowBitmap)
+		c.scratch.buffers[i][1] = memory.NewBufferBytes(values)
 
 		// Create the data from the buffers. It might be surprising that we don't
 		// set a type or a null count, but these fields are not used in the way that
@@ -251,7 +250,11 @@ func (c *ArrowBatchConverter) ArrowToBatch(data []*array.Data, b coldata.Batch) 
 					// corresponds.
 					bytes = make([]byte, 0)
 				}
-				coldata.BytesFromArrowSerializationFormat(vec.Bytes(), bytes, bytesArr.ValueOffsets())
+				offsets := bytesArr.ValueOffsets()
+				vecArr := vec.Bytes()
+				for i := 0; i < len(offsets)-1; i++ {
+					vecArr.Set(i, bytes[offsets[i]:offsets[i+1]])
+				}
 				arr = bytesArr
 			default:
 				panic(fmt.Sprintf("unexpected type %s", typ))

--- a/pkg/sql/logictest/testdata/logic_test/vectorize
+++ b/pkg/sql/logictest/testdata/logic_test/vectorize
@@ -675,3 +675,10 @@ SELECT a FROM t_39827 ORDER BY a LIMIT 2
 ----
 a
 foo
+
+# Regression test for #40227, an issue with flat bytes implementation.
+statement ok
+CREATE TABLE t_40227 AS SELECT g FROM generate_series(0, 5) AS g
+
+statement ok
+SELECT '' FROM t_40227 AS t1 JOIN t_40227 AS t2 ON true

